lib/compress/zstd_compress.c

```diff
@@ -5645,8 +5644,29 @@ size_t ZSTD_compressStream2( ZSTD_CCtx* cctx,
 
     /* transparent initialization stage */
     if (cctx->streamStage == zcss_init) {
-        FORWARD_IF_ERROR(ZSTD_CCtx_init_compressStream2(cctx, endOp, input->size), "CompressStream2 initialization failed");
-        ZSTD_setBufferExpectations(cctx, output, input);    /* Set initial buffer expectations now that we've initialized */
+        size_t const inputSize = input->size - input->pos;  /* no obligation to start from pos==0 */
+        size_t const totalInputSize = (cctx->savedInPosPlusOne == 0) ? inputSize : input->size - (cctx->savedInPosPlusOne - 1);
+        if ( (cctx->requestedParams.inBufferMode == ZSTD_bm_stable) /* input is presumed stable, across invocations */
+          && (endOp == ZSTD_e_continue)                             /* no flush requested, more input to come */
+          && (totalInputSize < ZSTD_BLOCKSIZE_MAX) ) {              /* not even reached one block yet */
```

Comment:
Once one of these conditions is false, do we need to reset `input->pos` back to `cctx->expectedInBuffer.pos`?

Does this code work with `ZSTD_compressStream2(cctx, output, input, ZSTD_e_flush)` & `ZSTD_compressStream2(cctx, output, input, ZSTD_e_end)`? It looks buggy to me, or I am not understanding what is going on?

Either way, could you please add a test case using `ZSTD_compressStream2()`?

Reply:
This code is run every time `ZSTD_compressStream2()` is invoked.
The specific part within the `if()` condition only triggers when `ZSTD_c_stableInBuffer` is set to `1`,
and there is a whole section within `zstreamtest` dealing with this scenario, starting line 830,
including `ZSTD_compressStream2(cctx, output, input, ZSTD_e_end)`.
They currently pass properly.

We can certainly consider adding more tests if these tests are not enough.
I'm trying to figure out which scenario isn't already tested and would fail with this code ?

Reply:
> I'm trying to figure out which scenario isn't already tested and would fail with this code ?

```c
ZSTD_inBuffer in = {src, 0, 100};
ZSTD_compressStream2(cctx, out, in, ZSTD_e_continue);
// in.pos = 100 now
in.size = 200;
ZSTD_compressStream2(cctx, out, in, ZSTD_e_end);
```

I believe in that scenario, the first 100 bytes of the input will be skipped.

> and there is a whole section within zstreamtest dealing with this scenario, starting line 830,
> including ZSTD_compressStream2(cctx, output, input, ZSTD_e_end).

But I don't believe that there are any tests with `ZSTD_e_continue` & changing the buffer size.

The only tests with `ZSTD_e_continue` and `ZSTD_e_flush` are tests to make sure that they are disallowed. And after this change, they should work, not fail.

https://github.com/facebook/zstd/blob/70df5de1b2fc291a58c1b3199dde11b353ccad8f/tests/zstreamtest.c#L873-L889